Sdk internal metrics - BatchExportProcessorDroppedCount via ObservableCounter

vishweshbankwar opened the pull request:
Towards #3880 


## Changes
Address https://github.com/open-telemetry/opentelemetry-dotnet/pull/4312#pullrequestreview-1377793906

Adding `otel.dotnet.sdk.batchprocessor.dropped_count` metric with following dimensions

`batchprocessor.data.type` - Type of data that is being exported e.g. `Activity` or `LogRecord`.
`batchprocessor.exporter.name` - Name of the exporter e.g. `OtlpExporter`.

This PR shows an alternate approach to the one in https://github.com/open-telemetry/opentelemetry-dotnet/pull/4312. It uses `ObservableCounter` so that we can use existing `DroppedCount` on `BatchExportProcessor`. The overall idea is to add a callback to the list of `otel.dotnet.sdk.batchprocessor.dropped_count` callbacks everytime a new `BatchExportProcessor` is created. 

**Memory Management**
The callback will be removed during the `Dispose` of processor which will also address https://github.com/open-telemetry/opentelemetry-dotnet/pull/4289#discussion_r1134390550. Confirmed this behavior by taking memory snapshots during and after execution of following function

![MicrosoftTeams-image](https://user-images.githubusercontent.com/17655898/234377038-928d6be4-5641-49e5-a908-30ed7bb4078f.png)

Here is the diff
![MicrosoftTeams-image (1)](https://user-images.githubusercontent.com/17655898/234377205-e456a24e-3937-425b-a6ea-861473bf0e56.png)
This confirms that the instance is properly disposed and not held by `SdkHealthReporter`.

In case if we don't clean up the Callbacks during `dispose` the diff will look like this
![MicrosoftTeams-image (2)](https://user-images.githubusercontent.com/17655898/234378011-32d6f7b2-f423-45f9-822c-0e753fc7917c.png)
Meaning the object could not be collected due to an active reference

===============
**Example Output**
![image](https://user-images.githubusercontent.com/17655898/234719418-4e615a96-77ce-4168-ba15-87111b8f7e99.png)

**NOTE:**
Current changes does not consider a case of multiple instances of Providers and BatchExportProcessors of same data type. This can be handled separately once we have a basic working version.


## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

reyang commented on src/OpenTelemetry/BatchExportProcessor.cs:
```diff
@@ -72,6 +77,11 @@ public abstract class BatchExportProcessor<T> : BaseExportProcessor<T>
+
+            this.exporterName = exporter.GetType().Name;
+            this.exportedDataType = typeof(T).Name;
+            this.processorId = Interlocked.Increment(ref processorCount);
```
What does this id mean? (e.g. how would the user use it?)

vishweshbankwar replied:
This is to be added as a dimension on metric (In a scenario where they have same exporterName and exportedDataType).

reyang reviewed (approved):
The code looks good to me, regarding the ID, if the intention is to distinguish different instances, maybe a GUID would serve the purpose better than an ordinal number (both would have unlimited cardinality and in practice should not be a big issue since we don't expect many instances from a single process).

vishweshbankwar commented:
> The code looks good to me, regarding the ID, if the intention is to distinguish different instances, maybe a GUID would serve the purpose better than an ordinal number (both would have unlimited cardinality and in practice should not be a big issue since we don't expect many instances from a single process).

@reyang Could you please elaborate on how GUID will be a better dimension than ordinal number?

reyang commented:
> @reyang Could you please elaborate on how GUID will be a better dimension than ordinal number?

For example if there are two processes sending the metrics and they both have ID = 0, would the user be confused.

vishweshbankwar commented:
> > @reyang Could you please elaborate on how GUID will be a better dimension than ordinal number?
> 
> For example if there are two processes sending the metrics and they both have ID = 0, would the user be confused.

Thanks!, Updated it to guid.

vishweshbankwar commented:
> > > @reyang Could you please elaborate on how GUID will be a better dimension than ordinal number?
> > 
> > 
> > For example if there are two processes sending the metrics and they both have ID = 0, would the user be confused.
> 
> Thanks!, Updated it to guid.

@reyang, I reverted it for now to integer. Added description for each dimension that can be used as reference for further discussion.

reyang commented:
> > > > @reyang Could you please elaborate on how GUID will be a better dimension than ordinal number?
> > > 
> > > 
> > > For example if there are two processes sending the metrics and they both have ID = 0, would the user be confused.
> > 
> > 
> > Thanks!, Updated it to guid.
> 
> @reyang, I reverted it for now to integer. Added description for each dimension that can be used as reference for further discussion.

👍

vishweshbankwar commented:
@open-telemetry/dotnet-maintainers  - Please have a look.
